File: hw/aimacana/u3/POO/java/rpg-system/src/main/java/ec/edu/espe/rpg/model/entities/Potion.java
package ec.edu.espe.rpg.model.entities;

import ec.edu.espe.rpg.model.interfaces.IConsumable;
import ec.edu.espe.rpg.model.interfaces.ISellable;

public class Potion extends Item implements IConsumable, ISellable {
    private double restorationAmount;

    public Potion(String id, String name, double weight, String description, double baseValue, double restorationAmount) {
        super(id, name, weight, description, baseValue);
        this.restorationAmount = restorationAmount;
    }

    public double getRestorationAmount() { return restorationAmount; }

    @Override
    public void consume(Character target) {
        System.out.println(target.getName() + " consumes " + getName());
        target.heal(restorationAmount);
    }

    @Override
    public double calculateSaleValue() {
        // Potions maintain their value unless consumed
        return getBaseValue();
    }
}

File: hw/aimacana/u3/POO/typescript/rpg-system/src/server/models/Potion.ts
import { Item } from './Item';
import type { IConsumable, ISellable } from './interfaces';
import type { Character } from './Character';

export class Potion extends Item implements IConsumable, ISellable {
  private restorationAmount: number;

  constructor(id: string, name: string, weight: number, description: string, baseValue: number, restorationAmount: number) {
    super(id, name, weight, description, baseValue);
    this.restorationAmount = restorationAmount;
  }

  public getRestorationAmount(): number { return this.restorationAmount; }

  public consume(target: Character): void {
    console.log(`${target.getName()} consumes ${this.getName()}`);
    target.heal(this.restorationAmount);
  }

  public calculateSaleValue(): number {
    // Potions maintain their value unless consumed
    return this.getBaseValue();
  }
}
